### DISCUSSION TITLE
Proposal: Customizable Indentation Display in Editor

### DISCUSSION THREAD

**Discussion by noahfraiture** [helix-editor/helix#11578] [Ideas]
# Summary

I propose a feature to allow customizable visual indentation in the editor. For instance, users could display 2-space indentation while maintaining 4-space indentation in the actual code. Of course it could be anything else but I'll stick to that example for the rest of this post. This would be purely a visual change, leaving the underlying code untouched.

# Motivation

While many languages recommend 4-space indentation for consistency, some developers prefer a more compact 2-space visual layout. This feature would allow developers to work with their preferred visual style without breaking standard indentation practices. The standard are very important when working with other people on a project and it's often impossible to choose another standard than the official one. In open source for instance it isn't feasible.

This feature would improve the editing experience by offering visual customization while preserving code consistency. Feedback and discussion are welcome!

**Comment by chtenb** [comment]
There is precedent in other text editors: https://marketplace.visualstudio.com/items?itemName=kylepaulsen.stretchy-spaces

**Comment by noahfraiture** [comment]
Is this something that would be accepted as a PR or is it something unwanted for helix ?

**Reply by pascalkuthe** [reply]
This would need some kind of general conceal virtual text system. That is planned eventually but doesn't exist yet and this PR is not the right place to flash that out. It would also likely need special handling since it would require that the cursor would be able to actually select parts of the concealed text instead of just expanding the selection if a cursor is inside.

I also feel this is too niche. You can achieve something similar with git hooks that replace the indentation on check-in/check-out

**Reply by chtenb** [reply]
How would the git hook work? If my understanding is correct, and a checkout-hook would changes spaces in the source code, wouldn't that show up as unstaged changes all the time?

**Reply by pascalkuthe** [reply]
I don't have an example but I know that git runs files trough the git filter pipeline before performing a diff. With that it should be possible to avoid this showing up as a change in a diff. Conceptually this is the same principle that the checkout as CLRF check-in as LF option uses (that one is just builtin because it's quite common).

**Reply by chtenb** [reply]
Interesting, thanks

**Comment by noahfraiture** [comment]
Thank you for the answer
